server/mem_agent/engine.py

```
@@ -0,0 +1,334 @@
+
+            def secure_open(file, *args, **kwargs):
+                """Open that restricts file access to allowed_path."""
+                # If file is a file object or path-like, get its string path
+                path = (
+                    file if isinstance(file, str) else getattr(file, "name", str(file))
+                )
+                full_path = os.path.abspath(path if path is not None else "")
+                if not full_path.startswith(allowed):
+                    raise PermissionError(
+                        f"Access to '{full_path}' is denied by sandbox."
+                    )
```

Ruff (0.14.1)

Abstract `raise` to an inner function

(TRY301)